### DISCUSSION TITLE
third-party licensing

### DISCUSSION THREAD

**Discussion by tkphd** [usnistgov/carpentries-development#3] [General]
A question came up in an email chain re. [LICENSE.md](/usnistgov/carpentries-development/LICENSE.md) and third-party source code:

> For other related licenses, they may have separate license files,  e.g. in subdirs correct?

My thoughts are: firstly, yes, that is one way to handle it. Secondly...

I believe that "best practice" for software is to include the full text of the license in the distribution of its source code, which is essentially what we do for our first-party code by reproducing the disclaimer of copyright and warranty in the top-level LICENSE file.

For third-party dependencies, you would be covered by retaining (or including, if absent) the full text of the license in the directory from which the dependency is included. I might have a source tree like the example below, where
* my code is in the top-level folders, with the NIST disclaimer of copyright in LICENSE.md
* the bottom of the top-level LICENSE.md file contains a notice that there are third-party libraries included, and an enumeration of which license applies to each dependency.
* specific libraries & licenses in the example are for flavor; discussion of including GPL'd code is tangential.

The dependencies would be neatly collected under `lib/third-party`, with a README providing some context and perhaps installation instructions for each. Then, within each third-party folder, you reproduce the original source code (with your modifications), including their LICENSE with original disclaimers and full-text legal code. It would be super duper to link to these legal documents from the top-level LICENSE.md, so there's full clarity on what's what.

I believe this to be necessary & sufficient to satisfy the terms of most open-source, header-only libraries in which you have made contributions and seek to redistribute as part of your GitHub repository. It may well mean messy build scripts and complex include statements, but *c'est la vie*.

```bash
$ tree -a my-awesome-code
my-awesome-code
├── .git
│   └── ...
├── CODEMETA.md
├── CODEOWNERS
├── LICENSE.md  # NIST disclaimer, followed by links to third-party LICENSEs under lib/third-party
├── README.md
├── bin
│   └── high-precision-imaging
├── docs
│   ├── Makefile
│   ├── my-awesome-code.md
│   └── my-awesome-code.pdf
├── lib
│   ├── myclass
│   │   ├── Makefile
│   │   ├── myclass.c
│   │   └── myclass.h
│   └── third-party
│       ├── README.md
│       ├── CImg
│       │   ├── LICENSE.md  # CeCILL-C full-text
│       │   ├── README.md
│       │   ├── bin
│       │   │   └── test
│       │   ├── include
│       │   │   └── CImg.h
│       │   ├── lib
│       │   └── src
│       │       ├── Makefile
│       │       └── main.c
│       └── mpfr
│           ├── LICENSE.md  # GPL full-text
│           ├── README.md
│           ├── bin
│           │   └── mpfr
│           ├── include
│           │   └── mpfr.h
│           ├── lib
│           └── src
│               ├── Makefile
│               ├── main.c
│               └── mpfr.c
└── src
    ├── analysis.py
    ├── myclass.py
    └── main.py

25 directories, 23 files
```

**Comment by ajnelson-nist** [comment]
Hi!  I'm coming here via the direction from [opensource-repo](https://github.com/usnistgov/opensource-repo).

I have a slightly different situation than the specific-licensing in this discussion thread.  With an external community where I frequently start code repositories, for code I start, my NIST public-domain licensing stamp goes into the files and the `LICENSE.md`.  However, the community has a blanket policy of using the Apache 2 license for their software resources.  What is the appropriate way to designate that?  I'd hope it would be, start with NIST's `LICENSE.md`, but replace this section:

```markdown
## Third-Party Software Licenses

This repository includes the following software, licensed by third
parties and not subject to the above disclaimer.

### *name of bundled dependency*

*Text of bundled dependency license*
```

with:

```
## Third-Party Software Licenses

This repository includes software not contributed by NIST,
licensed by third parties under the terms of the Apache 2
License and not subject to the above disclaimer.
 
(Copy-paste Apache 2 license file here.)
```

Somewhat along the same thread, those repositories currently happen to be written in mostly Python.  This is relevant because of a mechanism Python has to list the license as metadata.  [Here](https://github.com/casework/CASE-Utilities-Python/blob/0.3.0/setup.cfg#L13) is an example in one of the repositories I'm describing.  The classifier `License :: Public Domain` is used.

Once a code contribution comes in that is not-NIST, and agreed upon by the contributor to be subject to Apache 2, should the whole Python package be re-classified as an Apache 2 licensed repository? I.e. using this line: `License :: OSI Approved :: Apache Software License` (drawn from the PyPI [classifiers listing](https://pypi.org/classifiers/)).

**Reply by tkphd** [reply]
Good question! @guyer has thought a lot about third-party contributions, and I hope he'll weigh in.

We (The Govt) can't ask members of the public to waive copyright, nor can we reserve copyright (i.e., apply the Apache license) to our contributions, with limited exceptions. You're going to have a chimera of license terms, and that's fine, so long as you're clear about it. The original layout for "Third-Party Software Licenses" is intended to promote explicit enumeration of third-party contributions and the licensing terms applying to each. This works best at the folder or file level; I'm not sure if it's reasonable to reserve copyright on specific lines within files.

I suggest the following, which depends on providing a copy of Apache 2 in the top-level file `LICENSE-APACHE-2.md`:

```markdown
## Third-Party Software Licenses

The following contributions from members of the public are not subject to
the above disclaimer of copyright, but the disclaimer of warranty remains
in effect.

### Contributions under the Apache 2 license

The following files were contributed under the Apache 2 license, a full copy
of which is provided as [`LICENSE-APACHE-2.md`](/LICENSE-APACHE-2.md).

1. `filename.ext` by Name-of-Contributor (@their_handle)
```

Disclaimer: I am not a lawyer, and may be entirely mistaken.

**Reply by ajnelson-nist** [reply]
Thanks for the reply!  My situation is a blanket-special-case, where the external community has use of Apache 2 as a general practice.  Theoretically, this could mean not needing to go so far with enumerating files.

My community is also unlikely to do file-level enumerations.  Many of the issues would involve revising just a few lines within files.

Disclaimer: I am also not a lawyer.

**Reply by guyer** [reply]
I don't know that I have a lot to add to what @tkphd has already said.

NIST Counsel has approved dual licenses at times, and it may make sense to consult with them about this if this is going to be recurring. Ian Bell has asked for he and I to meet with Mark Madsen to discuss a dual-licensing situation; if it ever happens, I can ask about this, too.

Disclaimer: I was never even tempted to be a lawyer.
